Index: linux-2.6.21/fs/hfsplus/super.c
===================================================================
--- linux-2.6.21.orig/fs/hfsplus/super.c	2009-06-17 14:05:18.000000000 +0200
+++ linux-2.6.21/fs/hfsplus/super.c	2009-06-17 14:15:12.000000000 +0200
@@ -239,11 +239,13 @@
 		if (!hfsplus_parse_options(data, &sbi))
 			return -EINVAL;
 
+		if (vhdr->attributes & cpu_to_be32(HFSPLUS_VOL_JOURNALED))
+			printk("%s:%s[%d]\n", __FILE__, __func__, __LINE__);
 		if (!(vhdr->attributes & cpu_to_be32(HFSPLUS_VOL_UNMNT))) {
-			printk(KERN_WARNING "hfs: filesystem was not cleanly unmounted, "
-			       "running fsck.hfsplus is recommended.  leaving read-only.\n");
-			sb->s_flags |= MS_RDONLY;
-			*flags |= MS_RDONLY;
+//			printk(KERN_WARNING "hfs: filesystem was not cleanly unmounted, "
+//			       "running fsck.hfsplus is recommended.  leaving read-only.\n");
+//			sb->s_flags |= MS_RDONLY;
+//			*flags |= MS_RDONLY;
 		} else if (sbi.flags & HFSPLUS_SB_FORCE) {
 			/* nothing */
 		} else if (vhdr->attributes & cpu_to_be32(HFSPLUS_VOL_SOFTLOCK)) {
@@ -340,9 +342,9 @@
 	sb->s_maxbytes = MAX_LFS_FILESIZE;
 
 	if (!(vhdr->attributes & cpu_to_be32(HFSPLUS_VOL_UNMNT))) {
-		printk(KERN_WARNING "hfs: Filesystem was not cleanly unmounted, "
-		       "running fsck.hfsplus is recommended.  mounting read-only.\n");
-		sb->s_flags |= MS_RDONLY;
+	//	printk(KERN_WARNING "hfs: Filesystem was not cleanly unmounted, "
+	//	       "running fsck.hfsplus is recommended.  mounting read-only.\n");
+	//	sb->s_flags |= MS_RDONLY;
 	} else if (sbi->flags & HFSPLUS_SB_FORCE) {
 		/* nothing */
 	} else if (vhdr->attributes & cpu_to_be32(HFSPLUS_VOL_SOFTLOCK)) {
