Index: linux-2.6.21.7/ralink/wireless/Kconfig
===================================================================
--- linux-2.6.21.7.orig/ralink/wireless/Kconfig	2012-03-05 15:19:24.124766014 +0100
+++ linux-2.6.21.7/ralink/wireless/Kconfig	2012-03-05 15:20:19.240768370 +0100
@@ -2,6 +2,11 @@
 	tristate "Ralink RT2860 802.11n AP support"
 	depends on NET_RADIO 
 	
+config RT2860V2_AP_V24_DATA_STRUCTURE
+	bool
+	depends on RT2860V2_AP
+	default y
+	
 config  RT2860V2_AP_LED
 	bool "LED Support"
 	depends on RT2860V2_AP
@@ -22,10 +27,6 @@
 	bool "WMM ACM"
 	depends on RT2860V2_AP
 
-config  RT2860V2_AP_NINTENDO
-	bool "Nintendo"
-	depends on RT2860V2_AP
-
 config  RT2860V2_AP_MBSS
 	bool "MBSSID"
 	depends on RT2860V2_AP
@@ -64,10 +65,6 @@
 	bool "IDS (Intrusion Detection System) Support"
 	depends on RT2860V2_AP
 
-#config  RT2860V2_AP_MESH
-#	bool "MESH Support"
-#	depends on RT2860V2_AP
-
 config  RT2860V2_RT3XXX_AP_ANTENNA_DIVERSITY
 	bool "Antenna Diversity Support"
 	depends on RT2860V2_AP
@@ -84,10 +81,14 @@
 	bool "Memory Optimization"
 	depends on RT2860V2_AP
 
-#config  RT2860V2_AP_INTELLIGENT_RATE_ADAPTION
-#        bool "Intelligent Rate Adaption"
-#	depends on RT2860V2_AP
+config  RT2860V2_AP_VIDEO_TURBINE
+	bool "Video Turbine support"
+	depends on RT2860V2_AP
 
+config  RA_CLASSIFIER
+	tristate "Ralink Flow Classifier"
+	depends on RT2860V2_AP_VIDEO_TURBINE
+	default n
 
 #config  RT2860V2_EXT_CHANNEL_LIST
 #	bool "Extension Channel List"
