Index: linux-2.6.21/arch/mips/rt2880/Makefile
===================================================================
--- linux-2.6.21.orig/arch/mips/rt2880/Makefile	2009-05-22 16:10:26.000000000 +0200
+++ linux-2.6.21/arch/mips/rt2880/Makefile	2009-05-24 17:34:11.000000000 +0200
@@ -14,7 +14,7 @@
 	$(CC) $(CFLAGS) -c $< -o $*.o
 
 obj-y    := mipsIRQ.o reset.o init.o irq.o \
-            memory.o printf.o cmdline.o setup.o time.o 
+            memory.o printf.o cmdline.o setup.o time.o board.o
 obj-$(CONFIG_PCI) += pci.o
 obj-y += lm.o
 
Index: linux-2.6.21/arch/mips/rt2880/board.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.21/arch/mips/rt2880/board.c	2009-05-24 17:33:56.000000000 +0200
@@ -0,0 +1,41 @@
+#include <linux/autoconf.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/string.h>
+#include <linux/platform_device.h>
+#include <linux/kernel.h>
+
+static unsigned char ramips_eth_mac[6];
+
+static struct platform_device ramips_led = {
+	.id = 0,
+	.name = "rt3052-led",
+};
+
+static struct platform_device ramips_wdt = {
+	.id = 0,
+	.name = "ramips_wdt",
+};
+
+static struct platform_device ramips_eth =
+{
+	.id = 0,
+	.name = "ramips_eth",
+	.dev = {
+		.platform_data = ramips_eth_mac,
+	}
+};
+
+static __initdata struct platform_device *ramips_devs[3];
+
+int __init ramips_init_devices(void)
+{
+	int dev = 0;
+	ramips_devs[dev++] = &ramips_led;
+	ramips_devs[dev++] = &ramips_eth;
+	ramips_devs[dev++] = &ramips_wdt;
+	return platform_add_devices(ramips_devs, dev);
+}
+
+arch_initcall(ramips_init_devices);
