Index: linux-2.6.21/block/ll_rw_blk.c
===================================================================
--- linux-2.6.21.orig/block/ll_rw_blk.c	2009-05-22 14:23:36.000000000 +0200
+++ linux-2.6.21/block/ll_rw_blk.c	2009-05-22 14:25:49.000000000 +0200
@@ -3922,6 +3922,10 @@
 	return queue_var_show(max_hw_sectors_kb, (page));
 }
 
+static ssize_t queue_hw_sector_size_show(struct request_queue *q, char *page)
+{
+	return queue_var_show(q->hardsect_size, page);
+}
 
 static struct queue_sysfs_entry queue_requests_entry = {
 	.attr = {.name = "nr_requests", .mode = S_IRUGO | S_IWUSR },
@@ -3952,12 +3956,18 @@
 	.store = elv_iosched_store,
 };
 
+static struct queue_sysfs_entry queue_hw_sector_size_entry = {
+	.attr = {.name = "hw_sector_size", .mode = S_IRUGO },
+	.show = queue_hw_sector_size_show,
+};
+
 static struct attribute *default_attrs[] = {
 	&queue_requests_entry.attr,
 	&queue_ra_entry.attr,
 	&queue_max_hw_sectors_entry.attr,
 	&queue_max_sectors_entry.attr,
 	&queue_iosched_entry.attr,
+	&queue_hw_sector_size_entry.attr,
 	NULL,
 };
 
