Index: linux-2.6.21/drivers/usb/net/asix.c
===================================================================
--- linux-2.6.21.orig/drivers/usb/net/asix.c	2009-06-08 13:45:21.000000000 +0200
+++ linux-2.6.21/drivers/usb/net/asix.c	2009-06-08 15:45:13.000000000 +0200
@@ -1391,6 +1391,10 @@
 	USB_DEVICE (0x0b95, 0x1720),
 	.driver_info =  (unsigned long) &ax8817x_info,
 }, {
+	// VTech AT&T Plug&Share Ethernet USB 2.0 Adapter
+	USB_DEVICE (0x11eb, 0x0e12),
+	.driver_info =  (unsigned long) &ax8817x_info,
+}, {
 	// Hawking UF200, TrendNet TU2-ET100
 	USB_DEVICE (0x07b8, 0x420a),
 	.driver_info =  (unsigned long) &hawking_uf200_info,
